src/y0/algorithm/estimation.py

```diff
@@ -0,0 +1,334 @@
+    :returns: bool
+    """
+    if isinstance(treatments, list):
+        raise NotImplementedError(
+            "a-fixability on multiple treatments is an open research question"
+        )
+    descendants = graph.descendants_inclusive(treatments)
+    descendants_in_district = graph.get_district(treatments).intersection(descendants)
+    return 1 == len(descendants_in_district)
+
+
+def is_p_fixable(graph: NxMixedGraph, treatments: Union[Variable, List[Variable]]) -> bool:
+    """Check if the treatments are p-fixable.
+
+    This code was adapted from :mod:`ananke` ananke code at:
+    https://gitlab.com/causal/ananke/-/blob/dev/ananke/estimation/counterfactual_mean.py?ref_type=heads#L85-92
+
+    :param graph: A NxMixedGraph
+    :param treatments: A list of treatments
+    :raises NotImplementedError: p-fixability on multiple treatments is an open research question
+    :returns: bool
+    """
+    if isinstance(treatments, list):
+        raise NotImplementedError(
+            "p-fixability on multiple treatments is an open research question"
+        )
+    children = graph.directed.successors(treatments)
+    children_in_district = graph.get_district(treatments).intersection(children)
+    return 0 == len(children_in_district)
+
+
+def primal_ipw(
+    *,
+    data: pd.DataFrame,
+    graph: NxMixedGraph,
+    treatment: Variable,
+    treatment_value,
+    outcome: Variable,
+) -> float:
+    """Estimate the counterfactual mean E[Y(t)] with the Primal IPW estimator."""
+    beta_primal = get_beta_primal(
+        data=data,
+        graph=graph,
+        treatment=treatment,
+        treatment_value=treatment_value,
+        outcome=outcome,
+    )
+    return np.mean(beta_primal).item()
+
+
+def fit_binary_model(data, formula, weights=None) -> GLM:
+    """Fit a binary general linear model."""
+    return GLM.from_formula(
+        formula,
+        data=data,
+        family=Binomial(),
+        freq_weights=weights,
+    ).fit()
+
+
+def fit_continuous_glm(data, formula, weights=None) -> GLM:
+    """Fit a continuous general linear model."""
+    return GLM.from_formula(
+        formula,
+        data=data,
+        family=Gaussian(),
+        freq_weights=weights,
+    ).fit()
+
+
+def get_beta_primal(
+    *,
+    data: pd.DataFrame,
+    graph: NxMixedGraph,
+    treatment: Variable,
+    outcome: Variable,
+    treatment_value,
+    state_space_map: Dict[Variable, str],
+) -> np.array:
+    """Return the beta primal value for each row in the data.
+
+    This code was adapted from :mod:`ananke` ananke code at:
+    https://gitlab.com/causal/ananke/-/blob/dev/ananke/estimation/counterfactual_mean.py?ref_type=heads#L408-513
+
+    :param data: Given data
+    :param graph: A NxMixedGraph
+    :param treatment: Given treatment
+    :param outcome: Given outcome
+    :param treatment_value: Given treatment value
+    :param state_space_map: Contains the state space of a variable (binary/continuous)
```

Comment:
is there a helper function that takes a df and makes a state space map?

Reply:
There isn't one. I will work on it. Where do the helper functions reside?

Reply:
somewhere in estimation.py